[bug]/tapdb: proof storage clobber error for postgres backend

### Background

This failure has started to pop up on itests for the postgres backend: 
```
2024-06-12 16:42:33.793 [INF] GRDN: Received new proof file for asset ID 9bc313ea555c28227c86fa4a2b90f54dff46b14db4bbc56bd5d5e2e4dc3af34f, version=0,num_proofs=3
2024-06-12 16:42:33.807 [INF] FRTR: Marking parcel (txid=80bf98b4af11c552c95557806afa79468edec7b0a947f2c7b58306b0d4df4211) as confirmed!
2024-06-12 16:42:33.814 [ERR] FRTR: unable to transfer receiver proof: unable to log parcel delivery confirmation: failed to confirm transfer: unknown postgres error: ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)
```

It can be seen as a test failure in this itest run: https://github.com/lightninglabs/taproot-assets/actions/runs/9486428727/job/26140728303 
```
    --- FAIL: TestTaprootAssetsDaemon/multi_address (13.48s)
```

The issue appears to be with this query: https://github.com/lightninglabs/taproot-assets/blob/51e3512464e1efbed2abb6b3ea50f26fa41802f5/tapdb/sqlc/queries/assets.sql#L671-L691

If the CTE for `target_asset` returns more than one `asset_id`, then this issue will be triggered. AFAICT, sqlite will just take the very first asset ID. 

If we look at the table, we see that `asset_id` here is intended to be unique: https://github.com/lightninglabs/taproot-assets/blob/51e3512464e1efbed2abb6b3ea50f26fa41802f5/tapdb/sqlc/migrations/000002_assets.up.sql#L245-L256

This appears to be the root of the issue, as if you insert proofs for two diff script keys, but the same asset ID, then behavior is undefined across the two DB backends. However, this is only an issue for the db based proof file storage. The file based storage uses a full path of the outpoint+assetID+scriptKey (prev ID 3-tuple). The Universe based storage uses something similar. 

Most sub-systems end up using a `MultiArchive` that's based on some or all the proof storage backends, which my answer why this issue has been hidden for some time. 

### Suggested Fix

First we need to get a better feel of the impact. When we send out using the `ChainPorter`, the file based proofs are updated first, then later the DB proofs once proofs are delivered. 

Ultimately, I think we need a DB migration here to add the outpoint+script key as a composite 3-tuple primary key, similar to what we've done elsewhere. 



### Steps to reproduce

Run `make itest icase=multi_address` on postgres.

### Expected behavior

Should pass. 

### Actual behavior

Fails.

--- a/itest/addrs_test.go
+++ b/itest/addrs_test.go
@@ -603,7 +603,7 @@ func runMultiSendTest(ctxt context.Context, t *harnessTest, alice,
 	AssertAddrEvent(t.t, alice, aliceAddr2, 1, statusDetected)
 
 	// Mine a block to make sure the events are marked as confirmed.
-	_ = MineBlocks(t.t, t.lndHarness.Miner.Client, 1, 1)[0]
+	_ = MineBlocks(t.t, t.lndHarness.Miner.Client, 1, 1)
 
 	// Eventually the events should be marked as confirmed.
 	AssertAddrEventByStatus(t.t, bob, statusConfirmed, 2)

--- a/itest/assertions.go
+++ b/itest/assertions.go
@@ -818,7 +818,11 @@ func AssertAddrEventByStatus(t *testing.T, client taprpc.TaprootAssetsClient,
 			},
 		)
 		require.NoError(t, err)
-		require.Len(t, resp.Events, numEvents)
+
+		if len(resp.Events) != numEvents {
+			return fmt.Errorf("got %d events, wanted %d",
+				len(resp.Events), numEvents)
+		}
 
 		for _, event := range resp.Events {
 			if event.Status != filterStatus {
@@ -1410,7 +1414,7 @@ func AssertNumAssets(t *testing.T, ctx context.Context,
 	require.NoError(t, err)
 
 	// Ensure that the number of assets returned is correct.
-	require.Equal(t, numAssets, len(resp.Assets))
+	require.Len(t, resp.Assets, numAssets)
 
 	return resp
 }

--- a/tapdb/asset_minting.go
+++ b/tapdb/asset_minting.go
@@ -113,12 +113,13 @@ type (
 	// MintingBatchInit is used to create a new minting batch.
 	MintingBatchInit = sqlc.NewMintingBatchParams
 
-	// ProofUpdate is used to update a proof file on disk.
-	ProofUpdate = sqlc.UpsertAssetProofParams
-
 	// ProofUpdateByID is used to update a proof file on disk by asset
 	// database ID.
 	ProofUpdateByID = sqlc.UpsertAssetProofByIDParams
+
+	// FetchAssetID is used to fetch the primary key ID of an asset, by
+	// outpoint and tweaked script key.
+	FetchAssetID = sqlc.FetchAssetIDParams
 
 	// NewScriptKey wraps the params needed to insert a new script key on
 	// disk.
@@ -222,12 +223,14 @@ type PendingAssetStore interface {
 	FetchAssetsForBatch(ctx context.Context, rawKey []byte) ([]AssetSprout,
 		error)
 
-	// UpsertAssetProof inserts a new or updates an existing asset proof on
-	// disk.
-	//
-	// TODO(roasbeef): move somewhere else??
-	UpsertAssetProof(ctx context.Context,
-		arg sqlc.UpsertAssetProofParams) error
+	// FetchAssetID fetches the `asset_id` (primary key) from the assets
+	// table for a given asset identified by `Outpoint` and
+	// `TweakedScriptKey`.
+	FetchAssetID(ctx context.Context, arg FetchAssetID) ([]int64, error)
+
+	// UpsertAssetProofByID inserts a new or updates an existing asset
+	// proof on disk.
+	UpsertAssetProofByID(ctx context.Context, arg ProofUpdateByID) error
 
 	// FetchAssetMetaForAsset fetches the asset meta for a given asset.
 	FetchAssetMetaForAsset(ctx context.Context,
@@ -1644,9 +1647,28 @@ func (a *AssetMintingStore) MarkBatchConfirmed(ctx context.Context,
 		// As a final act, we'll now insert the proof files for each of
 		// the assets that were fully confirmed with this block.
 		for scriptKey, proofBlob := range mintingProofs {
-			err := q.UpsertAssetProof(ctx, ProofUpdate{
-				TweakedScriptKey: scriptKey.CopyBytes(),
-				ProofFile:        proofBlob,
+			// We need to fetch the table primary key `asset_id`
+			// first, as we need it to update the proof.
+			dbAssetIds, err := q.FetchAssetID(ctx, FetchAssetID{
+				TweakedScriptKey: scriptKey[:],
+			})
+			if err != nil {
+				return err
+			}
+
+			// We should not have more than one `asset_id`.
+			if len(dbAssetIds) > 1 {
+				return fmt.Errorf("expected 1 asset id, found "+
+					"%d with asset ids %v",
+					len(dbAssetIds), dbAssetIds)
+			}
+
+			// Upload proof by the dbAssetId, which is the _primary
+			// key_ of the asset in table assets, not the BIPS
+			// concept of `asset_id`.
+			err = q.UpsertAssetProofByID(ctx, ProofUpdateByID{
+				AssetID:   dbAssetIds[0],
+				ProofFile: proofBlob,
 			})
 			if err != nil {
 				return fmt.Errorf("unable to insert proof "+

--- a/tapdb/assets_store.go
+++ b/tapdb/assets_store.go
@@ -215,9 +215,10 @@ type ActiveAssetsStore interface {
 	UpsertManagedUTXO(ctx context.Context, arg RawManagedUTXO) (int64,
 		error)
 
-	// UpsertAssetProof inserts a new or updates an existing asset proof on
-	// disk.
-	UpsertAssetProof(ctx context.Context, arg ProofUpdate) error
+	// FetchAssetID fetches the `asset_id` (primary key) from the assets
+	// table for a given asset identified by `Outpoint` and
+	// `TweakedScriptKey`.
+	FetchAssetID(ctx context.Context, arg FetchAssetID) ([]int64, error)
 
 	// UpsertAssetProofByID inserts a new or updates an existing asset
 	// proof on disk.
@@ -1567,13 +1568,11 @@ func (a *AssetStore) importAssetFromProof(ctx context.Context,
 		return fmt.Errorf("unable to insert asset witness: %w", err)
 	}
 
-	// As a final step, we'll insert the proof file we used to generate all
-	// the above information.
-	scriptKeyBytes := newAsset.ScriptKey.PubKey.SerializeCompressed()
-	return db.UpsertAssetProof(ctx, ProofUpdate{
-		TweakedScriptKey: scriptKeyBytes,
-		Outpoint:         anchorPoint,
-		ProofFile:        proof.Blob,
+	// Upload proof by the dbAssetId, which is the _primary key_ of the
+	// asset in table assets, not the BIPS concept of `asset_id`.
+	return db.UpsertAssetProofByID(ctx, ProofUpdateByID{
+		AssetID:   assetIDs[0],
+		ProofFile: proof.Blob,
 	})
 }
 
@@ -1616,10 +1615,30 @@ func (a *AssetStore) upsertAssetProof(ctx context.Context,
 	// As a final step, we'll insert the proof file we used to generate all
 	// the above information.
 	scriptKeyBytes := proof.Asset.ScriptKey.PubKey.SerializeCompressed()
-	return db.UpsertAssetProof(ctx, ProofUpdate{
+
+	// We need to fetch the table primary key `asset_id` first, as we need
+	// it to update the proof. We could do this in one query, this gave
+	// issues with a postgresql backend. See:
+	// https://github.com/lightninglabs/taproot-assets/issues/951
+	dbAssetIds, err := db.FetchAssetID(ctx, FetchAssetID{
 		TweakedScriptKey: scriptKeyBytes,
 		Outpoint:         outpointBytes,
-		ProofFile:        proof.Blob,
+	})
+	if err != nil {
+		return err
+	}
+
+	// We should not have more than one `asset_id`.
+	if len(dbAssetIds) > 1 {
+		return fmt.Errorf("expected 1 asset id, found %d with asset "+
+			"ids %v", len(dbAssetIds), dbAssetIds)
+	}
+
+	// Upload proof by the dbAssetId, which is the _primary key_ of the
+	// asset in table assets, not the BIPS concept of `asset_id`.
+	return db.UpsertAssetProofByID(ctx, ProofUpdateByID{
+		AssetID:   dbAssetIds[0],
+		ProofFile: proof.Blob,
 	})
 }
 
@@ -1634,6 +1653,7 @@ func (a *AssetStore) ImportProofs(ctx context.Context, _ proof.HeaderVerifier,
 	proofs ...*proof.AnnotatedProof) error {
 
 	var writeTxOpts AssetStoreTxOptions
+
 	err := a.db.ExecTx(ctx, &writeTxOpts, func(q ActiveAssetsStore) error {
 		for _, p := range proofs {
 			if replace {
@@ -2675,6 +2695,7 @@ func (a *AssetStore) ConfirmParcelDelivery(ctx context.Context,
 		writeTxOpts    AssetStoreTxOptions
 		localProofKeys []asset.SerializedKey
 	)
+
 	err := a.db.ExecTx(ctx, &writeTxOpts, func(q ActiveAssetsStore) error {
 		// First, we'll fetch the asset transfer based on its outpoint
 		// bytes, so we can apply the delta it describes.
@@ -2826,13 +2847,14 @@ func (a *AssetStore) ConfirmParcelDelivery(ctx context.Context,
 			}
 			localProofKeys = append(localProofKeys, scriptKey)
 
-			// Now we can update the asset proof for the sender for
-			// this given delta.
-			err = q.UpsertAssetProof(ctx, ProofUpdate{
-				TweakedScriptKey: out.ScriptKeyBytes,
-				Outpoint:         out.AnchorOutpoint,
-				ProofFile:        receiverProof.Blob,
+			// Upload proof by the dbAssetId, which is the _primary
+			// key_ of the asset in table assets, not the BIPS
+			// concept of `asset_id`.
+			err = q.UpsertAssetProofByID(ctx, ProofUpdateByID{
+				AssetID:   newAssetID,
+				ProofFile: receiverProof.Blob,
 			})
+
 			if err != nil {
 				return err
 			}

--- a/tapdb/interfaces.go
+++ b/tapdb/interfaces.go
@@ -208,9 +208,9 @@ func (t *TransactionExecutor[Q]) ExecTx(ctx context.Context,
 	waitBeforeRetry := func(attemptNumber int) {
 		retryDelay := t.opts.randRetryDelay(attemptNumber)
 
-		log.Tracef("Retrying transaction due to tx serialization "+
-			"error, attempt_number=%v, delay=%v", attemptNumber,
-			retryDelay)
+		log.Tracef("Retrying transaction due to tx serialization or "+
+			"deadlock error, attempt_number=%v, delay=%v",
+			attemptNumber, retryDelay)
 
 		// Before we try again, we'll wait with a random backoff based
 		// on the retry delay.
@@ -222,7 +222,7 @@ func (t *TransactionExecutor[Q]) ExecTx(ctx context.Context,
 		tx, err := t.BatchedQuerier.BeginTx(ctx, txOptions)
 		if err != nil {
 			dbErr := MapSQLError(err)
-			if IsSerializationError(dbErr) {
+			if IsSerializationOrDeadlockError(dbErr) {
 				// Nothing to roll back here, since we didn't
 				// even get a transaction yet.
 				waitBeforeRetry(i)
@@ -240,7 +240,7 @@ func (t *TransactionExecutor[Q]) ExecTx(ctx context.Context,
 
 		if err := txBody(t.createQuery(tx)); err != nil {
 			dbErr := MapSQLError(err)
-			if IsSerializationError(dbErr) {
+			if IsSerializationOrDeadlockError(dbErr) {
 				// Roll back the transaction, then pop back up
 				// to try once again.
 				_ = tx.Rollback()
@@ -255,7 +255,7 @@ func (t *TransactionExecutor[Q]) ExecTx(ctx context.Context,
 		// Commit transaction.
 		if err = tx.Commit(); err != nil {
 			dbErr := MapSQLError(err)
-			if IsSerializationError(dbErr) {
+			if IsSerializationOrDeadlockError(dbErr) {
 				// Roll back the transaction, then pop back up
 				// to try once again.
 				_ = tx.Rollback()

--- a/tapdb/sqlc/assets.sql.go
+++ b/tapdb/sqlc/assets.sql.go
@@ -516,6 +516,48 @@ func (q *Queries) DeleteUTXOLease(ctx context.Context, outpoint []byte) error {
 	return err
 }
 
+const fetchAssetID = `-- name: FetchAssetID :many
+SELECT asset_id
+    FROM assets
+    JOIN script_keys 
+        ON assets.script_key_id = script_keys.script_key_id
+    JOIN managed_utxos utxos
+        ON assets.anchor_utxo_id = utxos.utxo_id
+    WHERE
+        (script_keys.tweaked_script_key = $1
+            OR $1 IS NULL)
+        AND (utxos.outpoint = $2
+            OR $2 IS NULL)
+`
+
+type FetchAssetIDParams struct {
+	TweakedScriptKey []byte
+	Outpoint         []byte
+}
+
+func (q *Queries) FetchAssetID(ctx context.Context, arg FetchAssetIDParams) ([]int64, error) {
+	rows, err := q.db.QueryContext(ctx, fetchAssetID, arg.TweakedScriptKey, arg.Outpoint)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var items []int64
+	for rows.Next() {
+		var asset_id int64
+		if err := rows.Scan(&asset_id); err != nil {
+			return nil, err
+		}
+		items = append(items, asset_id)
+	}
+	if err := rows.Close(); err != nil {
+		return nil, err
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+	return items, nil
+}
+
 const fetchAssetMeta = `-- name: FetchAssetMeta :one
 SELECT meta_data_hash, meta_data_blob, meta_data_type
 FROM assets_meta
@@ -2507,40 +2549,6 @@ func (q *Queries) UpsertAssetMeta(ctx context.Context, arg UpsertAssetMetaParams
 	var meta_id int64
 	err := row.Scan(&meta_id)
 	return meta_id, err
-}
-
-const upsertAssetProof = `-- name: UpsertAssetProof :exec
-WITH target_asset(asset_id) AS (
-    SELECT asset_id
-    FROM assets
-    JOIN script_keys 
-        ON assets.script_key_id = script_keys.script_key_id
-    JOIN managed_utxos utxos
-        ON assets.anchor_utxo_id = utxos.utxo_id
-    WHERE
-        (script_keys.tweaked_script_key = $2
-            OR $2 IS NULL)
-        AND (utxos.outpoint = $3
-            OR $3 IS NULL)
-)
-INSERT INTO asset_proofs (
-    asset_id, proof_file
-) VALUES (
-    (SELECT asset_id FROM target_asset), $1
-) ON CONFLICT (asset_id)
-    -- This is not a NOP, we always overwrite the proof with the new one.
-    DO UPDATE SET proof_file = EXCLUDED.proof_file
-`
-
-type UpsertAssetProofParams struct {
-	ProofFile        []byte
-	TweakedScriptKey []byte
-	Outpoint         []byte
-}
-
-func (q *Queries) UpsertAssetProof(ctx context.Context, arg UpsertAssetProofParams) error {
-	_, err := q.db.ExecContext(ctx, upsertAssetProof, arg.ProofFile, arg.TweakedScriptKey, arg.Outpoint)
-	return err
 }
 
 const upsertAssetProofByID = `-- name: UpsertAssetProofByID :exec

--- a/tapdb/sqlc/querier.go
+++ b/tapdb/sqlc/querier.go
@@ -44,6 +44,7 @@ type Querier interface {
 	FetchAddrEventByAddrKeyAndOutpoint(ctx context.Context, arg FetchAddrEventByAddrKeyAndOutpointParams) (FetchAddrEventByAddrKeyAndOutpointRow, error)
 	FetchAddrs(ctx context.Context, arg FetchAddrsParams) ([]FetchAddrsRow, error)
 	FetchAllNodes(ctx context.Context) ([]MssmtNode, error)
+	FetchAssetID(ctx context.Context, arg FetchAssetIDParams) ([]int64, error)
 	FetchAssetMeta(ctx context.Context, metaID int64) (FetchAssetMetaRow, error)
 	FetchAssetMetaByHash(ctx context.Context, metaDataHash []byte) (FetchAssetMetaByHashRow, error)
 	FetchAssetMetaForAsset(ctx context.Context, assetID []byte) (FetchAssetMetaForAssetRow, error)
@@ -157,7 +158,6 @@ type Querier interface {
 	UpsertAssetGroupKey(ctx context.Context, arg UpsertAssetGroupKeyParams) (int64, error)
 	UpsertAssetGroupWitness(ctx context.Context, arg UpsertAssetGroupWitnessParams) (int64, error)
 	UpsertAssetMeta(ctx context.Context, arg UpsertAssetMetaParams) (int64, error)
-	UpsertAssetProof(ctx context.Context, arg UpsertAssetProofParams) error
 	UpsertAssetProofByID(ctx context.Context, arg UpsertAssetProofByIDParams) error
 	UpsertAssetWitness(ctx context.Context, arg UpsertAssetWitnessParams) error
 	UpsertChainTx(ctx context.Context, arg UpsertChainTxParams) (int64, error)

--- a/tapdb/sqlc/queries/assets.sql
+++ b/tapdb/sqlc/queries/assets.sql
@@ -669,9 +669,8 @@ UPDATE chain_txns
 SET block_height = $2, block_hash = $3, tx_index = $4
 WHERE txn_id in (SELECT txn_id FROM target_txn);
 
--- name: UpsertAssetProof :exec
-WITH target_asset(asset_id) AS (
-    SELECT asset_id
+-- name: FetchAssetID :many
+SELECT asset_id
     FROM assets
     JOIN script_keys 
         ON assets.script_key_id = script_keys.script_key_id
@@ -681,15 +680,7 @@ WITH target_asset(asset_id) AS (
         (script_keys.tweaked_script_key = sqlc.narg('tweaked_script_key')
             OR sqlc.narg('tweaked_script_key') IS NULL)
         AND (utxos.outpoint = sqlc.narg('outpoint')
-            OR sqlc.narg('outpoint') IS NULL)
-)
-INSERT INTO asset_proofs (
-    asset_id, proof_file
-) VALUES (
-    (SELECT asset_id FROM target_asset), @proof_file
-) ON CONFLICT (asset_id)
-    -- This is not a NOP, we always overwrite the proof with the new one.
-    DO UPDATE SET proof_file = EXCLUDED.proof_file;
+            OR sqlc.narg('outpoint') IS NULL);
 
 -- name: UpsertAssetProofByID :exec
 INSERT INTO asset_proofs (

--- a/tapdb/sqlerrors.go
+++ b/tapdb/sqlerrors.go
@@ -53,6 +53,13 @@ func parseSqliteError(sqliteErr *sqlite.Error) error {
 			DbError: sqliteErr,
 		}
 
+	// A write operation could not continue because of a conflict within the
+	// same database connection.
+	case sqlite3.SQLITE_LOCKED:
+		return &ErrDeadlockError{
+			DbError: sqliteErr,
+		}
+
 	// Generic error, need to parse the message further.
 	case sqlite3.SQLITE_ERROR:
 		errMsg := sqliteErr.Error()
@@ -85,6 +92,13 @@ func parsePostgresError(pqErr *pgconn.PgError) error {
 	// Unable to serialize the transaction, so we'll need to try again.
 	case pgerrcode.SerializationFailure:
 		return &ErrSerializationError{
+			DbError: pqErr,
+		}
+
+	// A write operation could not continue because of a conflict within the
+	// same database connection.
+	case pgerrcode.DeadlockDetected:
+		return &ErrDeadlockError{
 			DbError: pqErr,
 		}
 
@@ -126,11 +140,39 @@ func (e ErrSerializationError) Error() string {
 	return e.DbError.Error()
 }
 
+// ErrDeadlockError is an error type which represents a database agnostic
+// error where transactions have led to cyclic dependencies in lock acquisition.
+type ErrDeadlockError struct {
+	DbError error
+}
+
+// Unwrap returns the wrapped error.
+func (e ErrDeadlockError) Unwrap() error {
+	return e.DbError
+}
+
+// Error returns the error message.
+func (e ErrDeadlockError) Error() string {
+	return e.DbError.Error()
+}
+
 // IsSerializationError returns true if the given error is a serialization
 // error.
 func IsSerializationError(err error) bool {
 	var serializationError *ErrSerializationError
 	return errors.As(err, &serializationError)
+}
+
+// IsDeadlockError returns true if the given error is a deadlock error.
+func IsDeadlockError(err error) bool {
+	var deadlockError *ErrDeadlockError
+	return errors.As(err, &deadlockError)
+}
+
+// IsSerializationOrDeadlockError returns true if the given error is either a
+// deadlock error or a serialization error.
+func IsSerializationOrDeadlockError(err error) bool {
+	return IsDeadlockError(err) || IsSerializationError(err)
 }
 
 // ErrSchemaError is an error type which represents a database agnostic error

